Search all SelectingItemsControl items with TextSearch on key input, not just realized ones

jonko0493 opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
This PR brings the behavior of `SelectingItemControl`s with `VirtualizingPanel` presenters into parity with those controls without virtualizing presenters.

## What is the current behavior?
Currently, `SelectingItemsControl`s with `VirtualizingPanel` item presenters (such as `ComboBox`) do not search the full contents of their boxes with text search, only the currently realized controls. This means ~that if a `ComboBox` is closed, typing does nothing, and even when opened,~ (this was slightly wrong!) it will only jump to items immediately in the popup.

## What is the updated/expected behavior with this PR?
Now, `ComboBox`es and other `SelectingItemsControl`s with virtualizing presenters will have proper `TextSearch` for their entire item sets.

## How was the solution implemented (if it's not obvious)?
The anonymous method that existed in the `SelectingItemsControl` class previously has been moved into a static method in `ItemsPresenter`. It's now used there and in `VirtualizingPanel` to check search their items/children for text search. (Note: I don't know if there is a preferred place for a static method like this; happy to move as appropriate.)

Unit test case was added to an existing test and the test was adapted so that ComboBoxes in tests will use a `VirtualizingStackPanel` like the real controls do to ensure the tests fail without this change.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
<!--- List any breaking changes here. -->

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
Fixes #12490.

jonko0493 committed:
Modify SelectingItemsControl to not just use unrealized items

jonko0493 committed:
Fixup null deref

jonko0493 committed:
Get unrealized items searched in comboboxes

jonko0493 committed:
Fixup one small comparison bug

jonko0493 committed:
Reset file that shouldn't have been changed

jonko0493 committed:
Try again

jonko0493 committed:
Revert frfr

jonko0493 committed:
Revert frfrfr

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

cla-avalonia commented:
@jonko0493,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

jonko0493 commented:
@cla-avalonia agree

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 commented:
Everything was passing prior to me merging master... not sure what changed, could be transient failures?

timunie commented:
@jonko0493 don't worry tests are a bit flaky atm

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 commented:
Hi folks! Is there something more I need to do to get this reviewed? It is needed for the application I'm developing and it's been a month since it was opened. Thanks!

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 commented:
@timunie I'm very sorry to ping you, but I'm wondering if there's anything else I need to do to get this reviewed. It's needed to make my project usable and I'd prefer not to take a dependency on the PR package. Please let me know if there's anything I can do!

timunie commented:
If you have paid support please ask for priority

jonko0493 commented:
@timunie I don't have paid support as we're a tiny open source project, so fair enough, I'll wait my turn. This PR has been open three months -- how long should I expect to wait before getting a review without paid support?

MrJul reviewed (changes requested):
Thank you for your contribution.

While the implementation has its merits, we now have different behaviors for standard panels and for `VirtualizingPanel`.
The non-virtualized implementation searches inside item containers (controls) directly. The new implementation for `VirtualizingPanel` searches only inside `Items`.

We should have the same implementation for both panels here, for consistency.
Looking at WPF, it always searches in `Items`, so I propose that we do the same, whatever the panel.

The search should look for:
 - `TextSearch.Text` (current)
 - `IContentControl.Content.ToString()` (current)
 - `item.ToString()`
 
(Ideally we'd also consider `DisplayMemberBinding`, like WPF does, but this is out of scope of this PR and can be added later).

jonko0493 commented:
@MikeCodesDotNET Thanks so much for letting me know and I'm sorry for misrepresenting how long the PR had been open -- that was truly a mistake on my part (I can't do date math well in my head, apparently lol).

@MrJul Thank you for the review! The changes you've suggested make good sense. I'll make them in the next few days.

jonko0493 committed:
Fixup per PR feedback

jonko0493 committed:
Remove documentation from internal method

jonko0493 committed:
Remove unused usings

jonko0493 commented:
@MrJul Sorry for the delay -- I am embarrassed that it took so long but life really hit me with a ton of stuff all at once right after I said I was going to do this. In any case, I believe I have addressed your feedback. If you could give this another look when you get a chance, I would greatly appreciate it!

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

jonko0493 committed:
Merge branch 'master' into UnrealizedItemsTextSearch

MrJul commented:
> @MrJul Sorry for the delay -- I am embarrassed that it took so long but life really hit me with a ton of stuff all at once right after I said I was going to do this.

No worries, life happens and in my opinion should always be the priority! :)
Thank you for coming back to this.

MrJul reviewed (approved):
LGTM, thank you!

jonko0493 commented:
@MrJul Thanks so much for your review and help!
